pkg/ddc/base/setup.go

```diff
@@ -29,90 +30,106 @@ func (b *TemplateEngine) Setup(ctx cruntime.ReconcileRequestContext) (ready bool
 	)
 
 	b.Log.Info("Setup the ddc engine", "runtime", ctx.Runtime)
-	// 1.Check if we should setup the master
-	// shouldSetupMaster, err
-	shouldSetupMaster, err = b.Implement.ShouldSetupMaster()
-	if err != nil {
-		return ready, err
-	}
-	if shouldSetupMaster {
-		err = b.Implement.SetupMaster()
+	// check if the dataset is ready
+	// if not ready, check if we should Setup the master, worker, ufs
+	// if ready, check if we should update ufs
+	if !utils.IsSetupDone(ctx.Dataset) {
+		// 1.Check if we should Setup1 the master
+		// shouldSetupMaster, err
+		shouldSetupMaster, err = b.Implement.ShouldSetupMaster()
 		if err != nil {
-			b.Log.Error(err, "SetupMaster")
-			return ready, err
+			return createready, updateready, err
+		}
+		if shouldSetupMaster {
+			err = b.Implement.SetupMaster()
+			if err != nil {
+				b.Log.Error(err, "SetupMaster")
+				return createready, updateready, err
+			}
 		}
-	}
-
-	// 2.Check if the master is ready, then go forward to workers setup
-	masterReady, err = b.Implement.CheckMasterReady()
-	if err != nil {
-		b.Log.Error(err, "Failed to check if it CheckMasterReady.")
-		return ready, err
-	}
 
-	if !masterReady {
-		return masterReady, err
-	}
+		// 2.Check if the master is ready, then go forward to workers Setup1
+		masterReady, err = b.Implement.CheckMasterReady()
+		if err != nil {
+			b.Log.Error(err, "Failed to check if it CheckMasterReady.")
+			return createready, updateready, err
+		}
 
-	shouldCheckUFS, err := b.Implement.ShouldCheckUFS()
-	if err != nil {
-		b.Log.Error(err, "Failed to check if it requires checking ufs.")
-		return ready, err
-	}
+		if !masterReady {
+			return createready, updateready, err
+		}
 
-	if shouldCheckUFS {
-		err = b.Implement.PrepareUFS()
+		shouldCheckUFS, err := b.Implement.ShouldCheckUFS()
 		if err != nil {
-			b.Log.Error(err, "Failed to prepare ufs.")
-			return ready, err
+			b.Log.Error(err, "Failed to check if it requires checking ufs.")
+			return createready, updateready, err
 		}
-	}
 
-	// 3.Check if we should setup the workers
-	shouldSetupWorkers, err = b.Implement.ShouldSetupWorkers()
-	if err != nil {
-		b.Log.Error(err, "Failed to check if it ShouldSetupWorkers.")
-		return ready, err
-	}
+		if shouldCheckUFS {
+			err = b.Implement.PrepareUFS()
+			if err != nil {
+				b.Log.Error(err, "Failed to prepare ufs.")
+				return createready, updateready, err
+			}
+		}
 
-	if shouldSetupWorkers {
-		err = b.Implement.SetupWorkers()
+		// 3.Check if we should Setup1 the workers
```

Comment:
Setup1 -> setup

Reply:
Thanks, just ignore it , we decided to seperate the process of update from setup, will commit it later.